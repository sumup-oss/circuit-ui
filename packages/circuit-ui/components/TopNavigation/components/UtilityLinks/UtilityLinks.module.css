.list {
  display: flex;
  align-items: center;
  padding: 0;
  margin: 0;
  list-style-type: none;
}

.item {
  height: 100%;
  border-left: var(--cui-border-width-kilo) solid var(--cui-border-divider);
}

.anchor {
  height: 100%;
  padding: 0 var(--cui-spacings-mega);
  text-decoration: none;
}

.icon {
  flex-shrink: 0;
  width: var(--cui-icon-sizes-mega);
  height: var(--cui-icon-sizes-mega);
}

@media (min-width: 480px) {
  .icon {
    margin-right: var(--cui-spacings-byte);
  }
}

@media (max-width: 479px) {
  .label {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    white-space: nowrap;
    border: 0;
    clip: rect(0 0 0 0);
  }
}

[aria-current="page"] .label {
  font-weight: var(--cui-font-weight-bold);
}
