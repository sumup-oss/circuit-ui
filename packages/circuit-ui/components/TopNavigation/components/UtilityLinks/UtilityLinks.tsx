/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MouseEvent, KeyboardEvent, FC, SVGProps, HTMLProps } from 'react';
import { css } from '@emotion/core';
import { Theme } from '@sumup/design-tokens';
import { Dispatch as TrackingProps } from '@sumup/collector';

import styled, { StyleProps } from '../../../../styles/styled';
import { hideVisually, navigationItem } from '../../../../styles/style-mixins';
import { useClickHandler } from '../../../../hooks/useClickHandler';
import Body from '../../../Body';

const anchorStyles = ({ theme }: StyleProps) => css`
  text-decoration: none;
  margin-right: ${theme.spacings.bit};
  padding: ${theme.spacings.byte};

  ${theme.mq.mega} {
    padding: ${theme.spacings.kilo} ${theme.spacings.mega};
  }
`;

const UtilityAnchor = styled.a(navigationItem, anchorStyles);

const iconStyles = (theme: Theme) => css`
  flex-shrink: 0;
  width: ${theme.iconSizes.mega};
  height: ${theme.iconSizes.mega};

  ${theme.mq.mega} {
    margin-right: ${theme.spacings.byte};
  }
`;

const labelStyles = ({ theme }: StyleProps) => css`
  ${theme.mq.untilMega} {
    ${hideVisually()};
  }
`;

const UtilityLabel = styled(Body)(labelStyles);

export interface UtilityLinkProps extends HTMLProps<HTMLAnchorElement> {
  /**
   * Display an icon in addition to the text to help to identify the link.
   * On narrow viewports, only the icon is displayed.
   */
  icon: FC<SVGProps<SVGSVGElement>>;
  /**
   * Short label to describe the target of the link.
   */
  label: string;
  /**
   * A valid path or URL to the link target.
   */
  href: string;
  /**
   * Function that's called when the link is clicked.
   */
  onClick?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * Whether the link is the currently active page.
   */
  isActive?: boolean;
  /**
   * Additional data that is dispatched with the tracking event.
   */
  tracking?: TrackingProps;
}

function UtilityLink({
  icon: Icon,
  label,
  onClick,
  tracking,
  ...props
}: UtilityLinkProps) {
  const handleClick = useClickHandler<MouseEvent | KeyboardEvent>(
    onClick,
    tracking,
    'utility-link',
  );

  return (
    <UtilityAnchor {...props} onClick={handleClick}>
      <Icon css={iconStyles} role="presentation" />
      <UtilityLabel
        size="one"
        variant={props.isActive ? 'highlight' : undefined}
      >
        {label}
      </UtilityLabel>
    </UtilityAnchor>
  );
}

const wrapperStyles = ({ theme }: StyleProps) => css`
  display: flex;
  align-items: center;

  ${theme.mq.untilMega} {
    &::after {
      content: '';
      display: block;
      height: 20px;
      width: ${theme.borderWidth.kilo};
      margin-right: ${theme.spacings.bit};
      background-color: ${theme.colors.n300};
    }
  }
`;

const UtilityLinksWrapper = styled.div(wrapperStyles);

export interface UtilityLinksProps {
  links: UtilityLinkProps[];
}

export function UtilityLinks({ links }: UtilityLinksProps): JSX.Element {
  return (
    <UtilityLinksWrapper>
      {links.map((link) => (
        <UtilityLink key={link.label} {...link} />
      ))}
    </UtilityLinksWrapper>
  );
}
