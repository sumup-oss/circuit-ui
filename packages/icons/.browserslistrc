# The browsers that are explicitly supported
chrome >= 63
firefox >= 67
safari >= 11.1
ios_saf >= 11.3
edge >= 79
opera >= 50
samsung >= 8.2

# The features that the script checks for
supports css-grid and supports object-entries and supports es6-module-dynamic-import
