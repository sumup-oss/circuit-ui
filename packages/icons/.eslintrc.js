module.exports = require('@sumup/foundry/eslint')({
  language: 'JavaScript',
  environments: ['Node'],
  frameworks: ['React', 'Jest'],
  openSource: true,
});
