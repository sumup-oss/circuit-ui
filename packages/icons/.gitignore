# Project #

node_modules
dist
__coverage__
__reports__
.npm
.eslintcache
.yarn-integrity
.env

# Packages #

# it's better to unpack these files and commit the raw source
# git has its own built in compression methods
*.7z
*.dmg
*.gz
*.iso
*.jar
*.rar
*.tar
*.zip

# Text editor files #

*.sublime-workspace
*.sublime-project
*.editorconfig
*.swp
.idea
.vscode

# Logs #

*.log

# OS generated files #

.DS_Store
.DS_Store?
._*
.Spotlight-V100
.Trashes
ehthumbs.db
Thumbs.db
