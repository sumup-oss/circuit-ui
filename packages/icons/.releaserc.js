module.exports = require('@sumup/foundry/semantic-release')({ publish: true });
