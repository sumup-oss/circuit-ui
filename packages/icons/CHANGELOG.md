Foundry uses [`semantic-release`](https://github.com/semantic-release/semantic-release) to automatically publish [GitHub release notes](https://github.com/sumup-oss/circuit-icons/releases) for each version.
