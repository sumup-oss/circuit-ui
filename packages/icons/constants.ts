/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import path from 'node:path';
import { fileURLToPath } from 'node:url';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore `import.meta` is fine
export const BASE_DIR = path.dirname(fileURLToPath(import.meta.url));
export const ICON_DIR = path.join(BASE_DIR, './web/v2');
export const DIST_DIR = path.join(BASE_DIR, 'dist');

export const SIZES = ['16', '24', '32'] as const;
export const CATEGORIES = [
  'Action',
  'Communication',
  'Payment method',
  'Card scheme',
  'Country flag',
  'Device',
  'File',
  'Finance',
  'Miscellaneous',
  'Navigation',
  'Notification',
  'Product and feature',
  'Social media',
  'Brand',
] as const;
