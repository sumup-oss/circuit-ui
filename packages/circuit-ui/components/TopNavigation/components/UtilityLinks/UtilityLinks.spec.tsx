/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { More } from '@sumup/icons';
import { KeyboardEvent, MouseEvent } from 'react';

import {
  axe,
  render,
  renderToHtml,
  userEvent,
} from '../../../../util/test-utils';

import { UtilityLinks } from './UtilityLinks';

describe('UtilityLinks', () => {
  const baseProps = {
    links: [
      {
        icon: More,
        label: 'More',
        href: '/more',
        onClick: jest.fn((event: MouseEvent | KeyboardEvent) => {
          event.preventDefault();
        }),
      },
    ],
  };

  describe('styles', () => {
    it('should match the snapshot', () => {
      const { container } = render(<UtilityLinks {...baseProps} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('business logic', () => {
    it('should call the onClick handler of a link', () => {
      const { getByText } = render(<UtilityLinks {...baseProps} />);

      const link = baseProps.links[0];

      userEvent.click(getByText(link.label));

      expect(link.onClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(<UtilityLinks {...baseProps} />);
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
