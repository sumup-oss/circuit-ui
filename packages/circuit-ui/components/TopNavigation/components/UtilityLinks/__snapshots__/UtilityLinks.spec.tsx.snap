// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UtilityLinks styles should match the snapshot 1`] = `
.circuit-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (max-width:767px) {
  .circuit-3::after {
    content: '';
    display: block;
    height: 20px;
    width: 1px;
    margin-right: 4px;
    background-color: #CCC;
  }
}

.circuit-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  outline: none;
  border-radius: 8px;
  color: #1A1A1A;
  background-color: #FFF;
  text-align: left;
  cursor: pointer;
  -webkit-transition: color 120ms ease-in-out,background-color 120ms ease-in-out;
  transition: color 120ms ease-in-out,background-color 120ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-right: 4px;
  padding: 8px;
}

.circuit-2:hover {
  background-color: #F5F5F5;
}

.circuit-2:active {
  background-color: #E6E6E6;
}

.circuit-2:focus {
  outline: 0;
  box-shadow: inset 0 0 0 4px #AFD0FE;
}

.circuit-2:focus::-moz-focus-inner {
  border: 0;
}

.circuit-2:disabled {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

@media (min-width:768px) {
  .circuit-2 {
    padding: 12px 16px;
  }
}

.circuit-0 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
}

@media (min-width:768px) {
  .circuit-0 {
    margin-right: 8px;
  }
}

.circuit-1 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 16px;
  line-height: 24px;
  margin-bottom: 0;
}

@media (max-width:767px) {
  .circuit-1 {
    border: 0;
    -webkit-clip: rect(0 0 0 0);
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    white-space: nowrap;
    width: 1px;
  }
}

<div>
  <div
    class="circuit-3"
  >
    <a
      class="circuit-2"
      href="/more"
    >
      <svg
        class="circuit-0"
        fill="none"
        height="24"
        role="presentation"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 21a9 9 0 100-18 9 9 0 000 18z"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
        <circle
          cx="12"
          cy="11.5"
          r=".5"
          stroke="currentColor"
          stroke-miterlimit="16"
        />
        <circle
          r="1"
          stroke="currentColor"
          stroke-miterlimit="16"
          transform="matrix(-1 0 0 1 12 12)"
        />
        <circle
          cx="16"
          cy="11.5"
          r=".5"
          stroke="currentColor"
          stroke-miterlimit="16"
        />
        <circle
          r="1"
          stroke="currentColor"
          stroke-miterlimit="16"
          transform="matrix(-1 0 0 1 16 12)"
        />
        <circle
          cx="8"
          cy="11.5"
          r=".5"
          stroke="currentColor"
          stroke-miterlimit="16"
        />
        <circle
          r="1"
          stroke="currentColor"
          stroke-miterlimit="16"
          transform="matrix(-1 0 0 1 8 12)"
        />
      </svg>
      <p
        class="circuit-1"
      >
        More
      </p>
    </a>
  </div>
</div>
`;
